### PR TITLE
Add Bosnian Road

### PR DESCRIPTION
**Pull request by Seren-35**
Genre info: https://yureklis.wordpress.com/2013/01/21/bosnian-road/

All original grids made it impossible for the loop to ever touch itself diagonally so no rules about it existed. Modern instances appear to always feature a rule against diagonal adjacency so this implementation adapts it too.

Existing clue notation is inconsistent. Every instance I've seen uses either black cells with white clues (somewhat confusing for a shading genre) or white cells with black clues (requires question marks for walls, which seem very common). I went with yajilin-like gray background with black clues and black outlines.

### CODE REVIEW

File: src-ui/list.html [hunk 1]
```diff
@@ -61,6 +61,7 @@ <h2 id="title"><span lang="ja">パズルの種類のリスト</span><span lang="
     <li data-pid="circlesquare"></li>
     <li data-pid="tetrochain"></li>
     <li data-pid="mrtile"></li>
+    <li data-pid="bosnianroad"></li>
```
Comment (x-sheep) [inline src-ui/list.html:64]: Could you file it under Loop genres instead?

File: src/variety/bosnianroad.js [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+			this.decodeCellQnum();
+			this.decodeCellAns();
+		},
+		encodeData: function() {
+			this.encodeCellQnum();
+			this.encodeCellAns();
+		}
+	},
+
+	AnsCheck: {
```
Comment (x-sheep) [inline src/variety/bosnianroad.js:234]: Could you add a 2x2 shaded check like in Snake? While it's technically redundant, in practice it's nicer to have a different error message stating that the loop goes back on itself.

**Review by x-sheep — CHANGES_REQUESTED** [review]
Thanks for the Pull Request! I only have a few small suggestions.

File: src/variety/bosnianroad.js (hunk 2 above)
Comment (Seren-35) [reply src/variety/bosnianroad.js:234]: It's somewhat underspecified but, as best as I can tell, unlike in Snake, a shaded 2x2 with nothing else in the grid is technically a valid Bosnian Road solution. It's admitted as such under the current implementation and I find it easier to argue for than against it. Perhaps I can make the 2x2 check only trigger if the connected component has more than 4 shaded cells?

Comment (x-sheep) [reply src/variety/bosnianroad.js:234]: > Perhaps I can make the 2x2 check only trigger if the connected component has more than 4 shaded cells?

That's a good idea.

File: src/variety/bosnianroad.js [hunk 3]
```diff
@@ -234,13 +234,47 @@
...
+				}
+
+				var bx = cell.bx,
+					by = cell.by;
+				var clist = bd.cellinside(bx, by, bx + 2, by + 2).filter(isShade);
+				if (clist.length < 4) {
+					continue;
+				}
+
+				var isPartOfBiggerLoop = clist.some(function(cell) {
```
Comment (x-sheep) [inline src/variety/bosnianroad.js:263]: This check could be simplified to measuring the size of the block the first cell is in.

Comment (Seren-35) [reply src/variety/bosnianroad.js]: 100% agreed, I just had a really hard time determining how to actually access that information and eventually gave up. Just got it though. This made reusing `check2x2Block` actually feasible - it performs a few redundant comparisons but I figure it's worth it for readability's sake.

**Comment by Seren-35** [comment]
Apologies for rewriting history, this repository contains mixed line endings and it makes my toolchain very unhappy.

**Review by x-sheep — APPROVED** [review]
Looks good to go. I'll release it tomorrow on pzprxs, since puzz.link updates are still suspended until @robx fixes the deployments.

**Pull request merged**